src/storage/mod.rs

```diff
@@ -1040,6 +1163,106 @@ impl Storage {
                         &Key::from_encoded(key),
                         limit,
                         &mut statistics,
+                        key_only,
+                    ).map_err(Error::from)
+                        .map(|r| {
+                            let mut valid_keys = 0;
+                            let mut bytes_read = 0;
+                            let mut stats = Statistics::default();
+                            r.iter().for_each(|r| {
+                                if let Ok(ref pair) = *r {
+                                    valid_keys += 1;
+                                    bytes_read += pair.0.len() + pair.1.len();
+                                }
+                            });
+                            stats.data.flow_stats.read_keys = valid_keys;
+                            stats.data.flow_stats.read_bytes = bytes_read;
+                            thread_ctx.collect_read_flow(ctx.get_region_id(), &stats);
+                            thread_ctx.collect_key_reads(CMD, valid_keys as u64);
+                            r
+                        });
+
+                    thread_ctx.collect_scan_count(CMD, &statistics);
+
+                    result
+                })
+                .then(move |r| {
+                    _timer.observe_duration();
+                    r
+                })
+        });
+
+        future::result(res)
+            .map_err(|_| Error::SchedTooBusy)
+            .flatten()
+    }
+
+    fn raw_batch_scan(
+        snapshot: Box<Snapshot>,
+        keys: Vec<Key>,
+        start_key_limit: usize,
+        total_limit: usize,
+        stats: &mut Statistics,
+        key_only: bool,
+    ) -> engine::Result<Vec<Result<KvPair>>> {
+        let mut cursor = snapshot.iter(IterOption::default(), ScanMode::Forward)?;
+        let mut pairs = vec![];
+        for key in &keys {
+            if !cursor.seek(key, &mut stats.data)? {
+                continue;
+            }
+            let mut key_pairs = 0;
+            while cursor.valid() && key_pairs < start_key_limit && pairs.len() < total_limit {
```

Comment:
Can we reuse the Storage::raw_scan function?

Reply:
If we decide there will be one total limit and split quota for each start key evenly, raw_scan can definitely be reused instead. So can you check the next comment about the limit.